// Author: Alex Thomson
//
// Protobuf type used as header for all caravan Messages.
syntax = "proto2";
import "scalar.proto";

message Header {
  // Sender's machine id.
  required uint64 from = 1;

  // Destination machine id.
  required uint64 to = 2;

  // Type of message.
  enum Type {
    RPC      = 1;  // New RPC request.
    CALLBACK = 2;  // Callback invocation from completed RPC.
    ACK      = 3;  // Acknowledgement of completed RPC request.
    DATA     = 4;  // Pure data, no function invocation (e.g. RPC result)
    SYSTEM   = 5;  // (NOT FOR USE BY APPS)
  }
  required Type type = 3;

  // Required for both RPC requests and CALLBACK invocations:
  optional string app = 11;
  optional string rpc = 12;

  // RPCs and callbacks may specify priority at which the requested procedure
  // should run.
  enum Priority {
    LOW = 0;
    HIGH = 1;
  }
  optional Priority priority = 13 [default = HIGH];

  // Optional for RPC requests (should NOT appear for CALLBACK invocations):
  //
  // RPC requests (but NOT callbacks) may request a response in one of
  // four ways (but a single RPC cannot request multiple responses):
  //
  //    (1) request a callback by setting both callback_app and callback_rpc
  optional string callback_app = 21;
  optional string callback_rpc = 22;
  //    (2) request an ack by setting ack_counter to a point to an atomic<int>
  optional uint64 ack_counter = 23;
  //    (3) request a data response by setting data_ptr to point to a
  //        MessageBuffer*
  optional uint64 data_ptr = 31;
  //    (4) request a data response by setting data_channel
  optional string data_channel = 32;

  // Stick miscellaneous stuff in here if you really want to.
  repeated bool   misc_bool   = 41;
  repeated uint64 misc_int    = 42;
  repeated bytes  misc_string = 43;
  repeated double misc_double = 44;
  repeated Scalar misc_scalar = 45;

  // EXTERNAL connection info. (NOT FOR USE BY APPS)
  optional string external_host = 51;
  optional int32 external_port = 52;
}

